pkg/ingester/user_tsdb.go

```diff
@@ -92,6 +92,7 @@ type ownedSeriesState struct {
 }
 
 type userTSDB struct {
+	hasDB          atomic.Bool
```

Comment:
I think the fact that there's an atomic flag that is only checked sometimes (but not in most of the accesses) can create more issues in the future, as now it's unclear whether the access to `u.db` has to be synchronized.

This problem is local to the fact that `blocksToDelete` is being passed as a callback to prometheus/tsdb, so I would suggest just fixing that callback to be properly synchronized, like this:
```diff
diff --git pkg/ingester/ingester.go pkg/ingester/ingester.go
index 13b66bd64c..de9e9b743e 100644
--- pkg/ingester/ingester.go
+++ pkg/ingester/ingester.go
@@ -30,6 +30,7 @@ import (
        "github.com/grafana/dskit/ring"
        "github.com/grafana/dskit/services"
        "github.com/grafana/dskit/tenant"
+       "github.com/oklog/ulid"
        "github.com/pkg/errors"
        "github.com/prometheus/client_golang/prometheus"
        "github.com/prometheus/client_golang/prometheus/promauto"
@@ -2666,6 +2667,14 @@ func (i *Ingester) createTSDB(userID string, walReplayConcurrency int) (*userTSD
        }
        userDB.triggerRecomputeOwnedSeries(recomputeOwnedSeriesReasonNewUser)

+       userDBHasDB := atomic.NewBool(false)
+       blocksToDelete := func(blocks []*tsdb.Block) map[ulid.ULID]struct{} {
+               if !userDBHasDB.Load() {
+                       return nil
+               }
+               return userDB.blocksToDelete(blocks)
+       }
+
        oooTW := i.limits.OutOfOrderTimeWindow(userID)
        // Create a new user database
        db, err := tsdb.Open(udir, userLogger, tsdbPromReg, &tsdb.Options{
@@ -2680,7 +2689,7 @@ func (i *Ingester) createTSDB(userID string, walReplayConcurrency int) (*userTSD
                WALSegmentSize:                        i.cfg.BlocksStorageConfig.TSDB.WALSegmentSizeBytes,
                WALReplayConcurrency:                  walReplayConcurrency,
                SeriesLifecycleCallback:               userDB,
-               BlocksToDelete:                        userDB.blocksToDelete,
+               BlocksToDelete:                        blocksToDelete,
                EnableExemplarStorage:                 true, // enable for everyone so we can raise the limit later
                MaxExemplars:                          int64(i.limiter.maxExemplarsPerUser(userID)),
                SeriesHashCache:                       i.seriesHashCache,
@@ -2721,6 +2730,7 @@ func (i *Ingester) createTSDB(userID string, walReplayConcurrency int) (*userTSD
        }

        userDB.db = db
+       userDBHasDB.Store(true)
        // We set the limiter here because we don't want to limit
        // series during WAL replay.
        userDB.limiter = i.limiter
```



I

Reply:
Thanks Oleg, applied your suggestion.